explorer/interpreter/type_checker.cpp

```diff
@@ -2335,10 +2336,28 @@ auto TypeChecker::MatchImpl(const InterfaceType& iface,
       *trace_stream_ << "matched with " << *impl.type << " as "
                      << *impl.interface << "\n\n";
     }
-    CARBON_ASSIGN_OR_RETURN(
-        const auto* subst_witness,
-        SubstituteCast<Witness>(*bindings_or_error, impl.witness));
-    return {subst_witness};
+
+    auto subst_witness_or_error =
+        SubstituteCast<Witness>(*bindings_or_error, impl.witness);
+
+    if (!subst_witness_or_error.ok()) {
+      std::string error;
+      llvm::raw_string_ostream stream(error);
+
+      stream << subst_witness_or_error.error().message() << "\n"
```

Comment:
It would definitely make more sense to have it near the member access error but I've had some trouble getting the `SourceLocation` at which instantiation was required. If @zygoloid has some suggestions I would be happy to implement them.